### DISCUSSION TITLE
How to create and use custom middleware?

### DISCUSSION THREAD

**Discussion by Zhuravlev1** [bukhalo/nestjs-telegraf#1023] [Q&A]
I am trying to implement custom middleware with a connection to DB using Prisma. My AppModule is:
```
@Module({
  imports: [
    TelegrafModule.forRootAsync({
      imports: [ConfigModule, LoggerModule],
      useFactory: (configService: ConfigService, logger: LoggerMiddleware) => {    
        return {
          token: configService.get<string>("TELEGRAM_TOKEN")!,
          middlewares: [sessionMiddleware, logger.use]
        };
      },
      inject: [ConfigService, LoggerMiddleware]
    }),
    PrismaModule
  ],
  controllers: [],
  providers: [...someProviders]
})
export class AppModule {} 
```

LoggerMiddleware:

```
@Injectable()
export class LoggerMiddleware {
  constructor(private readonly prisma: PrismaService) {}

  async use(ctx: Context, next: NextFunction) {
    const listUser = await this.prisma.user.findMany()
    console.log('listUser = ', listUser)
    next()
  }
}
```
LoggerModule:
```
@Module({
  imports: [PrismaModule],
  providers: [LoggerMiddleware, PrismaService],
  exports: [LoggerMiddleware]
})
export class LoggerModule {}
```

It starts without errors and code reaches my logger middleware but then I am getting an error:

> TypeError: Cannot read properties of undefined (reading 'prisma')

I have access to Prisma service from another provider and a connection to DB works. At the start, nest initializes all dependencies successfully and returns a result from db if call `logger.use({}, () => {})` like this:
```
useFactory:  async (configService: ConfigService, logger: LoggerMiddleware) => {   
        await logger.use({}, () => {}) // returns a result from DB on server start but failed then when code reaches middleware 
        return {
          token: configService.get<string>("TELEGRAM_TOKEN")!,
          middlewares: [sessionMiddleware, logger.use]
        };
      },
```
I don't understand why during execution it fell with this error. What did I do wrong? looks like my custom middleware doesn't have dependencies
<img width="1470" alt="image" src="https://user-images.githubusercontent.com/13619328/208265592-0806e18b-3170-4e03-933e-09d9e07d3a68.png">

**Comment by evilsprut** [comment]
Hey @Zhuravlev1, `middlewares` it just array of middleware functions in `Telegraf` module options. Inside the module, we simply take and call `bot.use()` for each element of the array. In this case, middleware has no Nest dependencies, because they exist outside Nest DI/Ioc.

**Reply by Zhuravlev1** [reply]
Thnaks! Could you give me some advice how to do it(create a middleware with dependencies)? In fact I try to add [telegraf-session-mongodb](https://github.com/alexnzarov/telegraf-session-mongodb) but i am not sure how to put connection.
I'd appreciate any help.

**Reply by evilsprut** [reply]
```typescript
import { Injectable, OnModuleInit } from '@nestjs/common';
import { InjectConnection } from '@nestjs/mongoose';
import { Connection } from 'mongoose';
import { session } from 'telegraf-session-mongodb';
import { InjectBot } from 'nestjs-telegraf'; 

@Injectable()
export class CatsService implements OnModuleInit {
  constructor(@InjectConnection() private connection: Connection, @InjectBot() private bot) {}

  onModuleInit() {
    this.bot.use(session(this.connection.db('main'), { collectionName: 'sessions' }))
  }
}
```
Something like that, not tested

**Reply by evilsprut** [reply]
Also you can get bot access in your `main.ts` file
```typescript
import { getBotToken } from 'nestjs-telegraf';
import { MongoClient } from 'mongodb';
import { session } from 'telegraf-session-mongodb';

const bot = app.get(getBotToken());

// code from telegram-session-mongodb, should work, but one more connection to DB not good
MongoClient.connect(process.env.MONGODB_URI, { useNewUrlParser: true, useUnifiedTopology: true })
  .then(client => {
    const db = client.db();
    bot.use(session(db, { collectionName: 'sessions' }));
  });
```

**Reply by evilsprut** [reply]
Also you can get access to bot in any place, use `getBotToken()` or `@InjectBot()`. improvise)
